File: 2170-count-number-of-maximum-bitwise-or-subsets/count-number-of-maximum-bitwise-or-subsets.java
class Solution {
    // Brute Force
    // int result = 0;
    // public void dfs(int index, int curresult, int[] nums, int max_or){
    //     if(index == nums.length){
    //         if(curresult == max_or){
    //             result += 1;
    //         }
    //         return;
    //     }
    //     dfs(index + 1, curresult, nums, max_or);
    //     dfs(index + 1, curresult | nums[index], nums, max_or);
    // }

    // public int countMaxOrSubsets(int[] nums) {
    //     int max_or = 0;
    //     for (int i: nums)
    //         max_or |= i;
        
    //     dfs(0, 0, nums, max_or);
    //     return result;
    // }

    // Brute Force without variable
    // public int dfs(int index, int curresult, int[] nums, int max_or){
    //     if(index == nums.length){
    //         return curresult == max_or ? 1 : 0;
    //     }
    //     return dfs(index + 1, curresult, nums, max_or) + dfs(index + 1, curresult | nums[index], nums, max_or);
    // }

    // public int countMaxOrSubsets(int[] nums) {
    //     int max_or = 0;
    //     for (int i: nums)
    //         max_or |= i;
    //     return dfs(0, 0, nums, max_or);
    // }

    //Memoization
    public int dfs(int index, int curresult, int[] nums, int max_or, HashMap<String, Integer> cache){
        String key = String.valueOf(index) + "_" + String.valueOf(curresult);
        if(cache.containsKey(key)) return cache.get(key);
        if(index == nums.length) return curresult == max_or ? 1 : 0;
        
        cache.put(key,  dfs(index + 1, curresult, nums, max_or, cache) + dfs(index + 1, curresult | nums[index], nums, max_or, cache));
        return cache.get(key);
    }

    public int countMaxOrSubsets(int[] nums) {
        HashMap<String, Integer> cache = new HashMap<String, Integer>();
        int max_or = 0;
        for (int i: nums)
            max_or |= i;
        return dfs(0, 0, nums, max_or, cache);
    }
}

File: 2170-count-number-of-maximum-bitwise-or-subsets/count-number-of-maximum-bitwise-or-subsets.py
class Solution:
    def countMaxOrSubsets(self, nums: List[int]) -> int:
        # # Brute Force
        # max_or = 0
        # for i in nums:
        #     max_or |= i
        # result = 0

        # def dfs(i, curr):
        #     nonlocal result
        #     if i  == len(nums): 
        #         if curr == max_or:
        #             result += 1
        #         return
        #     dfs(i + 1, curr)
        #     dfs(i + 1, curr | nums[i])
        # dfs(0, 0)
        # return result 

        # Brute Force with no variable
        max_or = 0
        for i in nums:
            max_or |= i

        def dfs(i, curr):
            if i  == len(nums): 
                return 1 if curr == max_or else 0
            return dfs(i + 1, curr) + dfs(i + 1, curr | nums[i])
        
        return dfs(0, 0) 